### PR TITLE
#2329, urgent mail support

### PR DESCRIPTION
**Pull request by punkstorm28**
A quick addition for urgent mail support, can be toggled from the
button in the overflow menu while composing,
When the message is received by the app on the other side, multiple
vibrations distinct from the normal is observed and the background of
lock screen notification is red.
For GSOC.

### CODE REVIEW

**Comment by cketti** [comment]
It's always a good idea to check the diff view that is displayed before creating a pull request. This pull request adds some new files that aren't used anywhere. Please clean that up.

**Comment by punkstorm28** [comment]
Hello,
I have cleaned up the files and committed,
Is it okay now?

**Comment by philipwhiuk** [comment]
I started commenting, but I've stopped because it's basically the same comments everywhere. Please follow the project code style. A short list of the ones I saw.

* Unnecessary new lines (often in files you otherwise haven't changed)
* `if (`
* `if` and `else` brackets should be on the same line
* Resource names `like_this` not `LikeThis`
* Remove the unused imports added

**Comment by punkstorm28** [comment]
@philipwhiuk ,I have cleaned the code to follow the project code style.
I will be more careful on the subsequent commits.
Thank you

**Comment by philipwhiuk** [comment]
By `if (` I meant that `if (` is what we expect.

And else blocks should look like:

```
if (something) {

} else {

}
```

You've also re-ordered imports - the correct order is:

```
import java

import android.

import com.fsck.

import timber.
```

Arguments should be like:

`somefunct(arg1, arg2, arg3)`

You are missing spaces in places.

**Comment by punkstorm28** [comment]
Hello,
I reformatted the code using idea's reformat feature and I re-ordered the imports.
Hope, it's better now.
Thank you

**Comment by Valodim** [comment]
More than half of your PR is still made up of unrelated formatting changes :(

**Comment by punkstorm28** [comment]
@Valodim 
Sorry, I hope I have cleaned up this time.

**Comment by punkstorm28** [comment]
Hi,
This was actually my first PR into open source, I now realized how important it is to keep things clean,
In the last commit I re-cloned the project, and made only the relevant changes wherever possible taking care of the code style.
Thank you

File: k9mail/src/main/java/com/fsck/k9/controller/MessagingController.java [hunk 1]
```diff
@@ -104,6 +104,7 @@
 import com.fsck.k9.search.SearchAccount;
 import com.fsck.k9.search.SearchSpecification;
 import com.fsck.k9.search.SqlQueryBuilder;
+
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/controller/MessagingController.java]: random unnecessary change

File: k9mail/src/main/java/com/fsck/k9/controller/MessagingControllerPushReceiver.java [hunk 2]
```diff
@@ -1,21 +1,21 @@
...
-import timber.log.Timber;
 
 import com.fsck.k9.Account;
 import com.fsck.k9.K9;
-import com.fsck.k9.mail.power.TracingPowerManager.TracingWakeLock;
 import com.fsck.k9.mail.Folder;
-
 import com.fsck.k9.mail.Message;
 import com.fsck.k9.mail.PushReceiver;
+import com.fsck.k9.mail.power.TracingPowerManager.TracingWakeLock;
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/controller/MessagingControllerPushReceiver.java]: unnecessary changes in otherwise untouched file. did you check the PR's diff?

File: k9mail/src/main/java/com/fsck/k9/message/MessageBuilder.java [hunk 3]
```diff
@@ -111,6 +112,9 @@ private void buildHeader(MimeMessage message) throws MessagingException {
             message.setHeader("User-Agent", context.getString(R.string.message_header_mua));
         }
 
+        if (isHighPriority) {
+            message.setHeader("X-Priority", context.getString(R.string.x_priority));
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/message/MessageBuilder.java]: this is a protocol mandated string, it should be a string constant in code.

File: k9mail/src/main/java/com/fsck/k9/notification/BaseNotifications.java [hunk 4]
```diff
@@ -12,7 +12,6 @@
 import com.fsck.k9.K9.NotificationQuickDelete;
 import com.fsck.k9.R;
 
-
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/notification/BaseNotifications.java]: unnecessary change

File: k9mail/src/main/java/com/fsck/k9/notification/NotificationController.java [hunk 5]
```diff
@@ -102,7 +103,13 @@ public void clearFetchingMailNotification(Account account) {
     }
 
     public void addNewMailNotification(Account account, LocalMessage message, int previousUnreadMessageCount) {
-        newMailNotifications.addNewMailNotification(account, message, previousUnreadMessageCount);
+        boolean hasHighPriority = false;
+
+        for (String iter : message.getHeader("X-Priority")) {
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/notification/NotificationController.java]: this logic should be part of `Message`. it's also unnecessary to check all headers, just the first one should suffice.

File: k9mail/src/main/java/com/fsck/k9/notification/NewMailNotifications.java [hunk 6]
```diff
@@ -59,9 +60,12 @@ public void addNewMailNotification(Account account, LocalMessage message, int un
                 int notificationId = result.getNotificationId();
                 cancelNotification(notificationId);
             }
-
-            createStackedNotification(account, result.getNotificationHolder());
-            createSummaryNotification(account, notificationData, false);
+            if (!isUrgent) {
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/notification/NewMailNotifications.java]: might as well be the positive case first, saves an inversion

File: k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java [hunk 7]
```diff
@@ -965,6 +968,17 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
                 break;
+            case R.id.is_urgent:
+                if (!isHighPriority) {
+                    item.setTitle("Normal");
+                    this.isHighPriority = true;
+                    setTitle(getTitle() + ":U");
+                } else {
+                    item.setTitle("Urgent");
+                    this.isHighPriority = false;
+                    setTitle(getTitle().toString().split(":")[0]);
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: this seems very volatile. I'm also not sure if the ":U" is necessary, it's not very obvious what it means and probably has little informational value to the user.

File: k9mail/src/main/java/com/fsck/k9/notification/DeviceNotifications.java [hunk 8]
```diff
@@ -42,6 +43,46 @@ public static DeviceNotifications newInstance(NotificationController controller,
         return new DeviceNotifications(controller, actionCreator, lockScreenNotification, wearNotifications);
     }
 
+    public Notification buildUrgentNotification(Account account, NotificationData notificationData) {
+        int unreadMessageCount = notificationData.getUnreadMessageCount();
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/notification/DeviceNotifications.java]: method seems ok to me, but I'm not involved in the notifications much. ping @cketti please review

**Comment by punkstorm28** [comment]
Hello,
I will redo it today,
Thank you

File: k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java [hunk 9]
```diff
@@ -747,6 +749,18 @@ private void onDiscard() {
...
 
+    private void onHighPriority(MenuItem item) {
+        if (isHighPriority) {
+            item.setTitle(getString(R.string.priority_urgent));
+            this.isHighPriority = false;
+            setTitle(getTitle().toString().split(",")[0]);
+        } else {
+            item.setTitle(getString(R.string.priority_normal));
+            this.isHighPriority = true;
+            setTitle(getTitle()+","+getString(R.string.priority_urgent));
```
Comment (punkstorm28) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: Changed from :U to ,Urgent

File: k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java [hunk 10]
```diff
@@ -750,6 +749,18 @@ private void onDiscard() {
...
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: Please use:

`getString(R.string.priority_urgent, getTitle());`

and make use of formatting arguments.

File: k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java [hunk 11]
```diff
@@ -750,6 +749,18 @@ private void onDiscard() {
         }
     }
 
+    private void onHighPriority(MenuItem item) {
+        if (isHighPriority) {
+            item.setTitle(getString(R.string.priority_urgent));
+            this.isHighPriority = false;
+            setTitle(getTitle().toString().split(",")[0]);
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: Please use a formatting string i.e.:

`setTitle(getString(R.string.high_priority, getTitle().toString().split(",")[0]))`

so that translation can handle it better.

File: k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java [hunk 12]
```diff
@@ -969,15 +980,7 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
-                if (!isHighPriority) {
-                    item.setTitle("Normal");
-                    this.isHighPriority = true;
-                    setTitle(getTitle() + ":U");
-                } else {
-                    item.setTitle("Urgent");
-                    this.isHighPriority = false;
-                    setTitle(getTitle().toString().split(":")[0]);
-                }
+                onHighPriority(item);
```
Comment (philipwhiuk) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: Please use consistent language - either 'High Priority' or 'Urgent'

File: k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java [hunk 13]
```diff
@@ -747,6 +749,18 @@ private void onDiscard() {
         }
     }
 
+    private void onUrgent(MenuItem item) {
```
Comment (punkstorm28) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: @philipwhiuk , Does this look okay? 
and I have changed all the isHighPriority to isUrgent.

**Comment by philipwhiuk** [comment]
That seems backwards. It's the X-Priority flag.

**Comment by punkstorm28** [comment]
@philipwhiuk , Sorry 
didn't get it, is it about the LocalMessage field naming, the urgentTag?

**Comment by philipwhiuk** [comment]
Given the MIME field is X-Priority it seems to me to be more logical to use 'High Priority' everywhere, not introduce a new word called Urgent.

**Comment by punkstorm28** [comment]
@philipwhiuk , I have updated the names with HighPriority :)

File: k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java [hunk 14]
```diff
@@ -728,6 +729,7 @@ public void performSendAfterChecks() {
         currentMessageBuilder = createMessageBuilder(false);
         if (currentMessageBuilder != null) {
             changesMadeSinceLastSave = false;
+            currentMessageBuilder.setHighPriority(isHighPriority);
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: this should be in createMessageBuilder

File: k9mail/src/main/java/com/fsck/k9/mailstore/LocalMessage.java [hunk 15]
```diff
@@ -43,6 +43,8 @@
     private String mimeType;
     private PreviewType previewType;
     private boolean headerNeedsUpdating = false;
+    private final String xPriorityContent = "priority_high";
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/mailstore/LocalMessage.java]: this can be a constant (= static final)

File: k9mail/src/main/java/com/fsck/k9/mailstore/LocalMessage.java [hunk 16]
```diff
@@ -173,6 +175,11 @@ public void setSubject(String subject) {
         headerNeedsUpdating = true;
     }
 
+    public boolean isHighPriority() {
+        if (getHeader(highPriorityTag).length>0 && getHeader(highPriorityTag)[0].equals(xPriorityContent))
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/mailstore/LocalMessage.java]: this can be in `MimeMessage`, and can use `getFirstHeader` there. also, code style violations

File: k9mail/src/main/java/com/fsck/k9/message/MessageBuilder.java [hunk 17]
```diff
@@ -111,6 +113,9 @@ private void buildHeader(MimeMessage message) throws MessagingException {
             message.setHeader("User-Agent", context.getString(R.string.message_header_mua));
         }
 
+        if (isHighPriority) {
+            message.setHeader(highPriorityHeader, context.getString(R.string.x_priority));
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/message/MessageBuilder.java]: the value can be a constant in code, no point having that in xml

File: k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java [hunk 18]
```diff
@@ -194,6 +194,7 @@
     private Action action;
 
     private boolean requestReadReceipt = false;
+    private boolean isHighPriority = false;
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: I feel like this belongs in `RecipientPresenter` (which should really be `MetadataPresenter`), but I suppose having it next to requestReadReceipt makes sense as well. feel free to leave as is.

File: k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java [hunk 19]
```diff
@@ -747,6 +749,18 @@ private void onDiscard() {
...
     }
 
+    private void onHighPriority(MenuItem item) {
+        if (isHighPriority) {
+            item.setTitle(getString(R.string.priority_high));
+            setTitle(getString(R.string.compose_title_compose));
+            this.isHighPriority = false;
+        } else {
+            item.setTitle(getString(R.string.priority_normal));
+            setTitle(String.format("%s,%s", getTitle(), getString(R.string.priority_high)));
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: I still don't like `setTitle` depending on `getTitle`. Just set it to another string resource

File: k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java (hunk 18 above)
Comment (punkstorm28) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: @Valodim , Thank you,
I have tried to address the points in this commit. :)

File: k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java [hunk 20]
```diff
@@ -747,6 +749,18 @@ private void onDiscard() {
...
     }
 
+    private void onHighPriority(MenuItem item) {
+        if (isHighPriority) {
+            item.setTitle(getString(R.string.priority_high));
+            setTitle(getString(R.string.compose_title_compose));
+            this.isHighPriority = false;
+        } else {
+            item.setTitle(getString(R.string.priority_normal));
+            setTitle(String.format("%s,%s", getString(R.string.compose_title_compose), getString(R.string.priority_high)));
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: why even do the format now :\

File: k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeMessage.java [hunk 21]
```diff
@@ -470,6 +473,13 @@ public void setCharset(String charset) throws MessagingException {
         }
     }
 
+    public boolean isHighPriority() {
+        if (getFirstHeader(highPriorityTag)!=null && getFirstHeader(highPriorityTag).equals(xPriorityContent)) {
```
Comment (Valodim) [inline k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeMessage.java]: you can check if `xPriorityContent.equals(tag)`, which will also return false if the header is null

File: k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeMessage.java [hunk 22]
```diff
@@ -60,6 +60,9 @@
     protected int mSize;
     private String serverExtra;
 
+    private static final String highPriorityTag = "X-Priority";
```
Comment (Valodim) [inline k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeMessage.java]: code style, constants are NAMED_LIKE_THIS

File: k9mail/src/main/java/com/fsck/k9/message/MessageBuilder.java [hunk 23]
```diff
@@ -42,6 +42,8 @@
     private final Context context;
     private final MessageIdGenerator messageIdGenerator;
     private final BoundaryGenerator boundaryGenerator;
+    private static final String highPriorityTag = "X-Priority";
```
Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/message/MessageBuilder.java]: this is duplicate now. I suggest moving it into `MimeHeader`, we already have a bunch of other headers there

File: k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeMessage.java (hunk 22 above)
Comment (punkstorm28) [inline k9mail-library/src/main/java/com/fsck/k9/mail/internet/MimeMessage.java]: :( , went unnoticed

File: k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java (hunk 20 above)
Comment (punkstorm28) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageCompose.java]: should I concat with a comma?

File: k9mail/src/main/java/com/fsck/k9/message/MessageBuilder.java (hunk 23 above)
Comment (punkstorm28) [inline k9mail/src/main/java/com/fsck/k9/message/MessageBuilder.java]: Ill do it now

**Comment by cketti** [comment]
Sorry, I should have chimed in earlier. But as I mentioned [here](https://github.com/k9mail/k-9/issues/2329#issuecomment-283898217), I don't want us to change the notification behavior based on a decision made by the sender of the message.

A visual indicator (I was thinking icon, rather than appended text) in notifications is fine. Changing the priority of notifications is not.

**Comment by punkstorm28** [comment]
@cketti 
urgentnotif
This is the notification, currently with LED blinks and extra vibrations,
should I disable the vibrations and LED blinks and let the red icon stay?

**Comment by cketti** [comment]
@punkstorm28: We use the account color as color for the notification. It'd be confusing if that was different with notifications for high priority messages. Even more so if an account is using red as account color.
For a visual indicator you could use an [ImageSpan](https://developer.android.com/reference/android/text/style/ImageSpan.html) to display a red exclamation mark icon in front of the subject.

Another thing: Please try to avoid pushing every single change to a branch there's an active pull request for. People who watch the project on GitHub will get a notification every time you do that.

**Comment by punkstorm28** [comment]
sure, Ill try today :)

**Comment by punkstorm28** [comment]
urgent
@cketti , something like this?
could also try making the whole subject have red font?

**Comment by philipwhiuk** [comment]
Rebased and improved in #2728

**Comment by philipwhiuk** [comment]
Thanks for your contribution by the way @punkstorm28  - you did a lot of the hard work here in terms of working out where everything should fit and making the first stab so 👍

**Comment by punkstorm28** [comment]
Thank you @philipwhiuk, means a lot.
